pkg/apply/applier_builder.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2021 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package apply
+
+import (
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/scheme"
+	"sigs.k8s.io/cli-utils/pkg/apply/info"
+	"sigs.k8s.io/cli-utils/pkg/apply/poller"
+	"sigs.k8s.io/cli-utils/pkg/apply/prune"
+	"sigs.k8s.io/cli-utils/pkg/inventory"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/polling"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ApplierBuilder struct {
+	// factory is only used to retrieve things that have not been provided explicitly.
+	factory                      util.Factory
+	invClient                    inventory.InventoryClient
+	client                       dynamic.Interface
+	discoClient                  discovery.CachedDiscoveryInterface
+	mapper                       meta.RESTMapper
+	restConfig                   *rest.Config
+	unstructuredClientForMapping func(*meta.RESTMapping) (resource.RESTClient, error)
+	statusPoller                 poller.Poller
+}
+
+// NewApplierBuilder returns a new ApplierBuilder.
+func NewApplierBuilder() *ApplierBuilder {
+	return &ApplierBuilder{
+		// Defaults, if any, go here.
+	}
+}
+
+func (b *ApplierBuilder) Build() (*Applier, error) {
+	bx, err := b.finalize()
+	if err != nil {
+		return nil, err
+	}
+	return &Applier{
+		pruner: &prune.Pruner{
+			InvClient: bx.invClient,
+			Client:    bx.client,
+			Mapper:    bx.mapper,
+		},
+		statusPoller:  bx.statusPoller,
+		invClient:     bx.invClient,
+		client:        bx.client,
+		openAPIGetter: bx.discoClient,
+		mapper:        bx.mapper,
+		infoHelper:    info.NewInfoHelper(bx.mapper, bx.unstructuredClientForMapping),
+	}, nil
+}
+
+func (b *ApplierBuilder) finalize() (*ApplierBuilder, error) {
```

Comment:
Do these all need to be finalized in the same function? Could we make a getter function for each option and just use those in the Build function to finalize each field individually?

Reply:
It can certainly be changed to work this way. I personally prefer the current style because I'm not sure what you are suggesting will look clearer in this particular case. I also like how you can put the cursor on a field and see where it's initialized and where it's used within a single method (to check that a field has been initialized before use).

If you don't feel strongly about it, I suggest we keep it this way.

Reply:
We are making a copy on the next line so that the builder is reusable. I thought you were talking about improving readability of the code.

Reply:
cool cool. i was also talking about readability, but that's not a deal breaker.